### PR TITLE
added ListGroupHooks issue #776

### PR DESCRIPTION
**Pull request by ericjaystevens**
Adds feature to list Group scoped web hooks.

### CODE REVIEW

File: group_hooks.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	PipelineEvents           bool       `url:"pipeline_events,omitempty"  json:"pipeline_events,omitempty"`
+	WikiPageEvents           bool       `url:"wiki_page_events,omitempty"  json:"wiki_page_events,omitempty"`
+	EnableSslVerification    bool       `url:"enable_ssl_verification,omitempty"  json:"enable_ssl_verification,omitempty"`
+	CreatedAt                *time.Time `url:"created_at,omitempty" json:"created_at,omitempty"`
+}
+
+//ListGroupHooks provides a list of Group scoped web hooks
+// GitLab API docs: https://docs.gitlab.com/ce/api/groups.html#list-group-hooks
+func (s *GroupsService) ListGroupHooks(gid interface{}) ([]*GroupHook, *Response, error) {
+
```
Comment (svanharmelen) [inline group_hooks.go]:
```suggestion
```

File: group_hooks.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	TagPushEvents            bool       `url:"tag_push_events,omitempty"  json:"tag_push_events,omitempty"`
+	NoteEvents               bool       `url:"note_events,omitempty"  json:"note_events,omitempty"`
+	JobEvents                bool       `url:"job_events,omitempty"  json:"job_events,omitempty"`
+	PipelineEvents           bool       `url:"pipeline_events,omitempty"  json:"pipeline_events,omitempty"`
+	WikiPageEvents           bool       `url:"wiki_page_events,omitempty"  json:"wiki_page_events,omitempty"`
+	EnableSslVerification    bool       `url:"enable_ssl_verification,omitempty"  json:"enable_ssl_verification,omitempty"`
+	CreatedAt                *time.Time `url:"created_at,omitempty" json:"created_at,omitempty"`
+}
+
+//ListGroupHooks provides a list of Group scoped web hooks
```
Comment (svanharmelen) [inline group_hooks.go]:
```suggestion
// ListGroupHooks gets a list of group hooks.
//
```

File: group_hooks.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	ConfidentialIssuesEvents bool       `url:"confidential_issues_events,omitempty"  json:"confidential_issues_events,omitempty"`
+	MergeRequestsEvents      bool       `url:"merge_requests_events,omitempty"  json:"merge_requests_events,omitempty"`
+	TagPushEvents            bool       `url:"tag_push_events,omitempty"  json:"tag_push_events,omitempty"`
+	NoteEvents               bool       `url:"note_events,omitempty"  json:"note_events,omitempty"`
+	JobEvents                bool       `url:"job_events,omitempty"  json:"job_events,omitempty"`
+	PipelineEvents           bool       `url:"pipeline_events,omitempty"  json:"pipeline_events,omitempty"`
+	WikiPageEvents           bool       `url:"wiki_page_events,omitempty"  json:"wiki_page_events,omitempty"`
+	EnableSslVerification    bool       `url:"enable_ssl_verification,omitempty"  json:"enable_ssl_verification,omitempty"`
+	CreatedAt                *time.Time `url:"created_at,omitempty" json:"created_at,omitempty"`
+}
```
Comment (svanharmelen) [inline group_hooks.go]: As this is a response struct, it doesn't need the `url` tags and the `omitempty` options here.

File: group_hooks.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+//
+
+package gitlab
+
+import (
+	"fmt"
+	"time"
+)
+
+// GroupHook represents a GitLab Group Hook
```
Comment (svanharmelen) [inline group_hooks.go]:
```suggestion
// GroupHook represents a GitLab group hook.
//
// GitLab API docs: https://docs.gitlab.com/ce/api/groups.html#list-group-hooks
```

File: group_hooks_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"testing"
+	"time"
+)
+
+func TestListGroupHooks(t *testing.T) {
+
```
Comment (svanharmelen) [inline group_hooks_test.go]:
```suggestion
```

File: group_hooks_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		EnableSslVerification:    true,
+		CreatedAt:                &datePointer,
+	}}
+
+	for i, groupHook := range groupHooks {
+		if groupHook.URL != want[i].URL {
+			t.Errorf("groups.ListGroupHooks.URL returned %s, want %s", groupHook.URL, want[i].URL)
+
+		}
+		i++
```
Comment (svanharmelen) [inline group_hooks_test.go]:
```suggestion
```

This is not needed as the `for` loop will set `i` to the current index each loop. Increasing `i` just before lopping again doesn't functionally do anything.

File: group_hooks_test.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+		PipelineEvents:           true,
+		WikiPageEvents:           true,
+		EnableSslVerification:    true,
+		CreatedAt:                &datePointer,
+	}}
+
+	for i, groupHook := range groupHooks {
+		if groupHook.URL != want[i].URL {
+			t.Errorf("groups.ListGroupHooks.URL returned %s, want %s", groupHook.URL, want[i].URL)
+
```
Comment (svanharmelen) [inline group_hooks_test.go]:
```suggestion
```

File: group_hooks_test.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  "wiki_page_events": true,
+  "enable_ssl_verification": true,
+  "created_at": "2012-10-12T17:04:47Z"
+			}
+		]`)
+	})
+
+	groupHooks, _, err := client.Groups.ListGroupHooks(1)
+	if err != nil {
+		log.Fatal(err)
```
Comment (svanharmelen) [inline group_hooks_test.go]:
```suggestion
		t.Error(err)
```

This logs the error and marks this test as failed, but continues the other tests.

File: group_hooks_test.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  "merge_requests_events": true,
+  "tag_push_events": true,
+  "note_events": true,
+  "job_events": true,
+  "pipeline_events": true,
+  "wiki_page_events": true,
+  "enable_ssl_verification": true,
+  "created_at": "2012-10-12T17:04:47Z"
+			}
+		]`)
```
Comment (svanharmelen) [inline group_hooks_test.go]:
```suggestion
		fmt.Fprint(w, `
[
	{
		"id": 1,
		"url": "http://example.com/hook",
		"group_id": 3,
		"push_events": true,
		"issues_events": true,
		"confidential_issues_events": true,
		"merge_requests_events": true,
		"tag_push_events": true,
		"note_events": true,
		"job_events": true,
		"pipeline_events": true,
		"wiki_page_events": true,
		"enable_ssl_verification": true,
		"created_at": "2012-10-12T17:04:47Z"
	}
]`)
```

File: group_hooks_test.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	})
+
+	groupHooks, _, err := client.Groups.ListGroupHooks(1)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	datePointer := time.Date(2012, 10, 12, 17, 4, 47, 0, time.UTC)
+	want := &[...]GroupHook{{
```
Comment (svanharmelen) [inline group_hooks_test.go]: Why would you take a pointer to an array here? Or why even use an array at all to be honest? It's much more idiomatic to just use a slice, so `[]Group`.

File: group_hooks_test.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+		JobEvents:                true,
+		PipelineEvents:           true,
+		WikiPageEvents:           true,
+		EnableSslVerification:    true,
+		CreatedAt:                &datePointer,
+	}}
+
+	for i, groupHook := range groupHooks {
+		if groupHook.URL != want[i].URL {
+			t.Errorf("groups.ListGroupHooks.URL returned %s, want %s", groupHook.URL, want[i].URL)
```
Comment (svanharmelen) [inline group_hooks_test.go]: Maybe you could do a `reflect.DeepEqual` on the whole struct here instead?

File: group_hooks_test.go (hunk 10 above)
Comment (ericjaystevens) [reply group_hooks_test.go]: Your right, A slice with reflect.DeepEqual would be a much better way to compare.

**Comment by svanharmelen** [comment]
It seems you processed some feedback, but not all yet. Let me know when I can review your PR again.

**Comment by ericjaystevens** [comment]
@svanharmelen Thanks for the great feedback, just pushed up my changes, should resolve everything you identified.

**Pull request merged**